File: Sources/MarketClock.swift
import Foundation

// ── 交易时段与节假日(智能刷新)────────────────────────────────────────────────
//
// 自选里所有市场都休市时,行情不会变,没必要每 30 秒请求一次:
// 拉取间隔放宽到"距最近一个开盘"(上限 30 分钟),开盘时刻准时恢复。
// 日历:美股按 NYSE 规则推算(含复活节、周末顺延、半日市);港股/A 股按官方公布的
// 休市表(每年更新一次,见下方数据来源)。表里没有的年份按"工作日即开市"处理——
// 宁可多拉几次,不会误判休市。另有实时兜底:行情时间戳显示几分钟内仍有成交,就按开市算。
// 各时段两端留几分钟余量,确保收盘价与开盘第一笔都能拿到。

enum MarketClock {

    struct Session: Equatable { let start: Int; let end: Int }   // 当地分钟数 [start, end)

    enum Exchange { case nyse, sse, hkex }

    static func exchange(for entry: WatchEntry) -> Exchange? {
        switch entry.market {
        case "crypto": return nil                                         // 7×24
        case "cn": return .sse
        case "hk": return .hkex
        default:
            // 期货/外汇(ES=F、EURUSD=X)近乎 24×5,加密后缀同理:按全天候处理
            if entry.symbol.contains("=") || entry.symbol.hasSuffix("-USD") { return nil }
            return .nyse
        }
    }

    static func timeZone(_ ex: Exchange) -> TimeZone {
        switch ex {
        case .nyse: return TimeZone(identifier: "America/New_York")!
        case .sse: return TimeZone(identifier: "Asia/Shanghai")!
        case .hkex: return TimeZone(identifier: "Asia/Hong_Kong")!
        }
    }

    // ── 官方休市表 ──
    // A 股:国务院办公厅 2025-11 放假通知(gov.cn content_7047091),与深交所交易日历接口逐日核对;
    //       2027 年安排通常 11 月公布,公布后补进来。
    // 港股:港交所 Trading Calendar and Holiday Schedule(2026-07-31 更新),含半日市。
    static let sseClosed: Set<String> = [
        "2026-01-01", "2026-01-02", "2026-02-16", "2026-02-17", "2026-02-18", "2026-02-19", "2026-02-20",
        "2026-02-23", "2026-04-06", "2026-05-01", "2026-05-04", "2026-05-05", "2026-06-19", "2026-09-25",
        "2026-10-01", "2026-10-02", "2026-10-05", "2026-10-06", "2026-10-07",
    ]
    static let hkexClosed: Set<String> = [
        "2026-01-01", "2026-02-17", "2026-02-18", "2026-02-19", "2026-04-03", "2026-04-06", "2026-04-07",
        "2026-05-01", "2026-05-25", "2026-06-19", "2026-07-01", "2026-10-01", "2026-10-19", "2026-12-25",
        "2027-01-01", "2027-02-08", "2027-02-09", "2027-03-26", "2027-03-29", "2027-04-05", "2027-05-13",
        "2027-06-09", "2027-07-01", "2027-09-16", "2027-10-01", "2027-10-08", "2027-12-27",
    ]
    static let hkexHalfDay: Set<String> = [   // 农历除夕、平安夜、除夕夜:只有上午
        "2026-02-16", "2026-12-24", "2026-12-31", "2027-02-05", "2027-12-24", "2027-12-31",
    ]

    // ── NYSE 规则(Rule 7.2)──

    private static func nthWeekday(_ n: Int, _ weekday: Int, month: Int, year: Int, cal: Calendar) -> DateComponents {
        var c = DateComponents(year: year, month: month, weekday: weekday, weekdayOrdinal: n)
        let d = cal.date(from: c)!
        c = cal.dateComponents([.year, .month, .day], from: d)
        return c
    }

    private static func lastMonday(of month: Int, year: Int, cal: Calendar) -> DateComponents {
        let c = DateComponents(year: year, month: month, weekday: 2, weekdayOrdinal: -1)
        return cal.dateComponents([.year, .month, .day], from: cal.date(from: c)!)
    }

    /// 复活节(格里历,匿名算法)
    static func easter(_ y: Int) -> (month: Int, day: Int) {
        let a = y % 19, b = y / 100, c = y % 100, d = b / 4, e = b % 4
        let f = (b + 8) / 25, g = (b - f + 1) / 3, h = (19 * a + b - d - g + 15) % 30
        let i = c / 4, k = c % 4, l = (32 + 2 * e + 2 * i - h - k) % 7
        let m = (a + 11 * h + 22 * l) / 451
        return ((h + l - 7 * m + 114) / 31, (h + l - 7 * m + 114) % 31 + 1)
    }

    static func nyseHolidays(_ year: Int) -> Set<String> {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = timeZone(.nyse)
        func key(_ c: DateComponents) -> String { String(format: "%04d-%02d-%02d", c.year!, c.month!, c.day!) }
        func key(_ d: Date) -> String { key(cal.dateComponents([.year, .month, .day], from: d)) }
        /// 固定日:周六提前到周五、周日顺延到周一;元旦逢周六不补(规则 7.2)
        func observed(_ month: Int, _ day: Int, saturdayRule: Bool = true) -> String? {
            let d = cal.date(from: DateComponents(year: year, month: month, day: day))!
            switch cal.component(.weekday, from: d) {
            case 7: return saturdayRule ? key(cal.date(byAdding: .day, value: -1, to: d)!) : nil
            case 1: return key(cal.date(byAdding: .day, value: 1, to: d)!)
            default: return key(d)
            }
        }
        var days = Set<String>()
        [observed(1, 1, saturdayRule: false), observed(6, 19), observed(7, 4), observed(12, 25)]
            .compactMap { $0 }.forEach { days.insert($0) }
        days.insert(key(nthWeekday(3, 2, month: 1, year: year, cal: cal)))    // MLK
        days.insert(key(nthWeekday(3, 2, month: 2, year: year, cal: cal)))    // Washington's Birthday
        days.insert(key(lastMonday(of: 5, year: year, cal: cal)))             // Memorial Day
        days.insert(key(nthWeekday(1, 2, month: 9, year: year, cal: cal)))   // Labor Day
        days.insert(key(nthWeekday(4, 5, month: 11, year: year, cal: cal)))   // Thanksgiving
        let e = easter(year)
        let goodFriday = cal.date(byAdding: .day, value: -2, to: cal.date(from: DateComponents(year: year, month: e.month, day: e.day))!)!
        days.insert(key(goodFriday))
        // 次年元旦逢周六:NYSE 不在本年 12/31 补休(规则如此),无需处理
        return days
    }

    /// NYSE 13:00 提前收盘:感恩节次日、平安夜、7/3(均须为工作日且非假日)
    static func nyseEarlyClose(_ year: Int) -> Set<String> {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = timeZone(.nyse)
        let holidays = nyseHolidays(year)
        func key(_ d: Date) -> String {
            let c = cal.dateComponents([.year, .month, .day], from: d)
            return String(format: "%04d-%02d-%02d", c.year!, c.month!, c.day!)
        }
        var days = Set<String>()
        let thanksgiving = cal.date(from: nthWeekday(4, 5, month: 11, year: year, cal: cal))!
        days.insert(key(cal.date(byAdding: .day, value: 1, to: thanksgiving)!))
        for (m, d) in [(12, 24), (7, 3)] {
            let date = cal.date(from: DateComponents(year: year, month: m, day: d))!
            let wd = cal.component(.weekday, from: date)
            if (2...6).contains(wd), !holidays.contains(key(date)) { days.insert(key(date)) }
        }
        return days
    }

    private static var nyseCache: [Int: (closed: Set<String>, early: Set<String>)] = [:]

    /// 某交易所某个当地日期的交易时段;空 = 休市
    static func sessions(_ ex: Exchange, year: Int, month: Int, day: Int, weekday: Int) -> [Session] {
        guard (2...6).contains(weekday) else { return [] }
        let key = String(format: "%04d-%02d-%02d", year, month, day)
        switch ex {
        case .nyse:
            let table = nyseCache[year] ?? (nyseHolidays(year), nyseEarlyClose(year))
            nyseCache[year] = table
            if table.closed.contains(key) { return [] }
            return [Session(start: 9 * 60 + 25, end: (table.early.contains(key) ? 13 : 16) * 60 + 5)]
        case .sse:
            if sseClosed.contains(key) { return [] }
            return [Session(start: 9 * 60 + 10, end: 11 * 60 + 35), Session(start: 12 * 60 + 55, end: 15 * 60 + 5)]
        case .hkex:
            if hkexClosed.contains(key) { return [] }
            // 港交所全日市含延长早市(12:00-13:00),9:30-16:00 连续;半日市只到 12:00
            return [Session(start: 9 * 60 + 15, end: hkexHalfDay.contains(key) ? 12 * 60 + 15 : 16 * 60 + 15)]
        }
    }

    private static func local(_ date: Date, _ ex: Exchange) -> (y: Int, m: Int, d: Int, wd: Int, minute: Int) {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = timeZone(ex)
        let c = cal.dateComponents([.year, .month, .day, .weekday, .hour, .minute], from: date)
        return (c.year!, c.month!, c.day!, c.weekday!, c.hour! * 60 + c.minute!)
    }

    static func isOpen(_ entry: WatchEntry, at date: Date) -> Bool {
        guard let ex = exchange(for: entry) else { return true }
        let t = local(date, ex)
        return sessions(ex, year: t.y, month: t.m, day: t.d, weekday: t.wd)
            .contains { t.minute >= $0.start && t.minute < $0.end }
    }

    /// 下一次任一时段开始的时刻(最长往后看 20 天,够跨过国庆长假)
    static func nextOpen(_ entry: WatchEntry, after date: Date) -> Date? {
        guard let ex = exchange(for: entry) else { return date }
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = timeZone(ex)
        let today = cal.startOfDay(for: date)
        for offset in 0...20 {
            guard let base = cal.date(byAdding: .day, value: offset, to: today) else { continue }
            let c = cal.dateComponents([.year, .month, .day, .weekday], from: base)
            for s in sessions(ex, year: c.year!, month: c.month!, day: c.day!, weekday: c.weekday!) {
                if let t = cal.date(byAdding: .minute, value: s.start, to: base), t > date { return t }
            }
        }
        return nil
    }

    /// 智能刷新间隔:任一市场开市(或刚有成交)→ 原间隔;全休市 → 等到最近开盘(上限 30 分钟)
    static func refreshInterval(_ entries: [WatchEntry], base: TimeInterval, now: Date,
                                lastTrade: [String: Date] = [:]) -> TimeInterval {
        guard !entries.isEmpty, !entries.contains(where: { isOpen($0, at: now) }) else { return base }
        // 实时兜底:日历说休市,但行情时间戳 5 分钟内还在走 → 日历可能漏了特殊交易日,按开市算
        if entries.contains(where: { e in lastTrade[e.symbol].map { now.timeIntervalSince($0) < 300 } ?? false }) {
            return base
        }
        let opens = entries.compactMap { nextOpen($0, after: now) }
        let untilOpen = opens.map { $0.timeIntervalSince(now) }.min() ?? 1800
        return min(1800, max(base, untilOpen + 5))
    }

    static func anyOpen(_ entries: [WatchEntry], at date: Date) -> Bool {
        entries.contains { isOpen($0, at: date) }
    }
}

/// 图表链接:美股/A 股/港股走 TradingView,指数/期货/外汇/加密走 Yahoo
func chartURL(for entry: WatchEntry) -> URL? {
    let s = entry.symbol
    let tvIndex: [String: String] = ["^GSPC": "SP:SPX", "^SPX": "SP:SPX", "^IXIC": "NASDAQ:IXIC",
                                      "^NDX": "NASDAQ:NDX", "^DJI": "DJ:DJI", "^VIX": "CBOE:VIX",
                                      "^RUT": "TVC:RUT", "^HSI": "HSI:HSI"]
    var tv: String?
    switch entry.market {
    case "cn":
        let sh = s.hasPrefix("6") || s.hasPrefix("5") || s.hasPrefix("9")
        tv = (sh ? "SSE:" : "SZSE:") + s
    case "hk":
        tv = "HKEX:" + String(Int(s) ?? 0)
    case "crypto":
        tv = nil
    default:
        if let mapped = tvIndex[s] { tv = mapped }
        else if !s.contains("^"), !s.contains("="), !s.hasSuffix("-USD") { tv = s }
    }
    if let tv, let q = tv.addingPercentEncoding(withAllowedCharacters: .alphanumerics.union(CharacterSet(charactersIn: ":.-"))) {
        return URL(string: "https://www.tradingview.com/chart/?symbol=\(q)")
    }
    let y = s.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? s
    return URL(string: "https://finance.yahoo.com/quote/\(y)")
}

File: Windows/Whirlpool.Core/MarketClock.cs
using System.Text.RegularExpressions;

namespace Whirlpool.Core;

/// <summary>
/// Exchange sessions and holidays for smart refresh (same data and rules as the macOS MarketClock).
/// NYSE holidays and 1:00 p.m. early closes are computed from the exchange rules; China A-share and
/// HKEX closures come from the official tables below and must be extended each year when published.
/// Years without data fall back to "weekdays are open", and a trade timestamp younger than five
/// minutes always counts as open.
/// </summary>
public static class MarketClock
{
    public enum Exchange { Nyse, Sse, Hkex }
    public readonly record struct Session(int Start, int End);   // local minutes [Start, End)

    // China A-shares: State Council notice (gov.cn content_7047091), checked day by day against the SZSE trading calendar.
    public static readonly HashSet<DateOnly> SseClosed = Dates(
        "2026-01-01", "2026-01-02", "2026-02-16", "2026-02-17", "2026-02-18", "2026-02-19", "2026-02-20",
        "2026-02-23", "2026-04-06", "2026-05-01", "2026-05-04", "2026-05-05", "2026-06-19", "2026-09-25",
        "2026-10-01", "2026-10-02", "2026-10-05", "2026-10-06", "2026-10-07");
    // HKEX Trading Calendar and Holiday Schedule (updated 2026-07-31).
    public static readonly HashSet<DateOnly> HkexClosed = Dates(
        "2026-01-01", "2026-02-17", "2026-02-18", "2026-02-19", "2026-04-03", "2026-04-06", "2026-04-07",
        "2026-05-01", "2026-05-25", "2026-06-19", "2026-07-01", "2026-10-01", "2026-10-19", "2026-12-25",
        "2027-01-01", "2027-02-08", "2027-02-09", "2027-03-26", "2027-03-29", "2027-04-05", "2027-05-13",
        "2027-06-09", "2027-07-01", "2027-09-16", "2027-10-01", "2027-10-08", "2027-12-27");
    public static readonly HashSet<DateOnly> HkexHalfDay = Dates(
        "2026-02-16", "2026-12-24", "2026-12-31", "2027-02-05", "2027-12-24", "2027-12-31");

    private static HashSet<DateOnly> Dates(params string[] values) => values.Select(v => DateOnly.Parse(v, System.Globalization.CultureInfo.InvariantCulture)).ToHashSet();

    public static Exchange? ExchangeFor(WatchEntry entry) => entry.Market switch
    {
        "crypto" => null,
        "cn" => Exchange.Sse,
        "hk" => Exchange.Hkex,
        _ => YahooExchange(entry.Symbol),
    };

    private static readonly HashSet<string> UsIndices = ["^GSPC", "^SPX", "^DJI", "^IXIC", "^NDX", "^RUT", "^VIX", "^NYA", "^XAX",
        "^SOX", "^OEX", "^DJT", "^DJU", "^W5000", "^TNX", "^TYX", "^FVX", "^IRX"];
    private static readonly HashSet<string> HongKongIndices = ["^HSI", "^HSCE"];

    /// <summary>
    /// Yahoo symbols name their venue: no suffix is a US listing, and .HK/.SS/.SZ use the tables above. Other venues
    /// (7203.T, VOD.L, ^N225…) have no modeled calendar, so they never count as closed.
    /// </summary>
    public static Exchange? YahooExchange(string symbol)
    {
        // Futures and FX (ES=F, EURUSD=X) trade almost 24×5; crypto pairs (BTC-USD, ETH-EUR) 24×7.
        if (symbol.Contains('=') || Regex.IsMatch(symbol, "-[A-Z]{3,}$")) return null;
        if (symbol.StartsWith('^')) return UsIndices.Contains(symbol) ? Exchange.Nyse : HongKongIndices.Contains(symbol) ? Exchange.Hkex : null;
        var dot = symbol.LastIndexOf('.');
        if (dot < 0) return Exchange.Nyse;
        return symbol[(dot + 1)..] switch { "HK" => Exchange.Hkex, "SS" or "SZ" => Exchange.Sse, _ => null };
    }

    private static TimeZoneInfo Zone(string iana, string windows)
    {
        try { return TimeZoneInfo.FindSystemTimeZoneById(iana); }
        catch (Exception e) when (e is TimeZoneNotFoundException or InvalidTimeZoneException) { return TimeZoneInfo.FindSystemTimeZoneById(windows); }
    }
    private static readonly TimeZoneInfo NewYork = Zone("America/New_York", "Eastern Standard Time");
    private static readonly TimeZoneInfo Shanghai = Zone("Asia/Shanghai", "China Standard Time");
    private static readonly TimeZoneInfo HongKong = Zone("Asia/Hong_Kong", "China Standard Time");
    public static TimeZoneInfo ZoneOf(Exchange ex) => ex switch { Exchange.Nyse => NewYork, Exchange.Sse => Shanghai, _ => HongKong };

    // ── NYSE Rule 7.2 ──
    public static (int Month, int Day) Easter(int y)
    {
        int a = y % 19, b = y / 100, c = y % 100, d = b / 4, e = b % 4;
        int f = (b + 8) / 25, g = (b - f + 1) / 3, h = (19 * a + b - d - g + 15) % 30;
        int i = c / 4, k = c % 4, l = (32 + 2 * e + 2 * i - h - k) % 7, m = (a + 11 * h + 22 * l) / 451;
        return ((h + l - 7 * m + 114) / 31, (h + l - 7 * m + 114) % 31 + 1);
    }
    private static DateOnly Nth(int n, DayOfWeek day, int month, int year)
    {
        var d = new DateOnly(year, month, 1);
        while (d.DayOfWeek != day) d = d.AddDays(1);
        return d.AddDays(7 * (n - 1));
    }
    private static DateOnly LastMonday(int month, int year)
    {
        var d = new DateOnly(year, month, DateTime.DaysInMonth(year, month));
        while (d.DayOfWeek != DayOfWeek.Monday) d = d.AddDays(-1);
        return d;
    }
    /// <summary>Saturday holidays move to Friday, Sunday to Monday; New Year's Day on a Saturday is not observed.</summary>
    private static DateOnly? Observed(int year, int month, int day, bool saturdayRule = true)
    {
        var d = new DateOnly(year, month, day);
        return d.DayOfWeek switch { DayOfWeek.Saturday => saturdayRule ? d.AddDays(-1) : null, DayOfWeek.Sunday => d.AddDays(1), _ => d };
    }
    public static HashSet<DateOnly> NyseHolidays(int year)
    {
        var days = new HashSet<DateOnly>();
        foreach (var d in new[] { Observed(year, 1, 1, false), Observed(year, 6, 19), Observed(year, 7, 4), Observed(year, 12, 25) })
            if (d is DateOnly value) days.Add(value);
        days.Add(Nth(3, DayOfWeek.Monday, 1, year));
        days.Add(Nth(3, DayOfWeek.Monday, 2, year));
        days.Add(LastMonday(5, year));
        days.Add(Nth(1, DayOfWeek.Monday, 9, year));
        days.Add(Nth(4, DayOfWeek.Thursday, 11, year));
        var (em, ed) = Easter(year);
        days.Add(new DateOnly(year, em, ed).AddDays(-2));
        return days;
    }
    /// <summary>1:00 p.m. closes: the day after Thanksgiving, Christmas Eve and July 3 when they are trading weekdays.</summary>
    public static HashSet<DateOnly> NyseEarlyClose(int year)
    {
        var holidays = NyseHolidays(year);
        var days = new HashSet<DateOnly> { Nth(4, DayOfWeek.Thursday, 11, year).AddDays(1) };
        foreach (var d in new[] { new DateOnly(year, 12, 24), new DateOnly(year, 7, 3) })
            if (d.DayOfWeek is not (DayOfWeek.Saturday or DayOfWeek.Sunday) && !holidays.Contains(d)) days.Add(d);
        return days;
    }
    private static readonly Dictionary<int, (HashSet<DateOnly> Closed, HashSet<DateOnly> Early)> NyseCache = [];

    public static IReadOnlyList<Session> Sessions(Exchange ex, DateOnly day)
    {
        if (day.DayOfWeek is DayOfWeek.Saturday or DayOfWeek.Sunday) return [];
        switch (ex)
        {
            case Exchange.Nyse:
                lock (NyseCache)
                {
                    if (!NyseCache.TryGetValue(day.Year, out var table)) NyseCache[day.Year] = table = (NyseHolidays(day.Year), NyseEarlyClose(day.Year));
                    if (table.Closed.Contains(day)) return [];
                    return [new(9 * 60 + 25, (table.Early.Contains(day) ? 13 : 16) * 60 + 5)];
                }
            case Exchange.Sse:
                return SseClosed.Contains(day) ? [] : [new(9 * 60 + 10, 11 * 60 + 35), new(12 * 60 + 55, 15 * 60 + 5)];
            default:
                // HKEX full days include the extended morning session (continuous 9:30–16:00); half days end at noon.
                return HkexClosed.Contains(day) ? [] : [new(9 * 60 + 15, HkexHalfDay.Contains(day) ? 12 * 60 + 15 : 16 * 60 + 15)];
        }
    }

    public static bool IsOpen(WatchEntry entry, DateTimeOffset now)
    {
        if (ExchangeFor(entry) is not Exchange ex) return true;
        var local = TimeZoneInfo.ConvertTime(now, ZoneOf(ex));
        var minute = local.Hour * 60 + local.Minute;
        return Sessions(ex, DateOnly.FromDateTime(local.DateTime)).Any(s => minute >= s.Start && minute < s.End);
    }

    public static DateTimeOffset? NextOpen(WatchEntry entry, DateTimeOffset after)
    {
        if (ExchangeFor(entry) is not Exchange ex) return after;
        var zone = ZoneOf(ex);
        var today = DateOnly.FromDateTime(TimeZoneInfo.ConvertTime(after, zone).DateTime);
        for (int offset = 0; offset <= 20; offset++)
        {
            var day = today.AddDays(offset);
            foreach (var s in Sessions(ex, day))
            {
                var local = day.ToDateTime(TimeOnly.MinValue).AddMinutes(s.Start);
                var start = new DateTimeOffset(local, zone.GetUtcOffset(local));
                if (start > after) return start;
            }
        }
        return null;
    }

    /// <summary>Base interval while any market is open (or just traded); otherwise until the next open, at most 30 minutes.</summary>
    public static double RefreshSeconds(IReadOnlyList<WatchEntry> entries, double baseSeconds, DateTimeOffset now,
                                        IReadOnlyDictionary<string, DateTimeOffset>? lastTrade = null)
    {
        if (entries.Count == 0 || entries.Any(e => IsOpen(e, now))) return baseSeconds;
        if (lastTrade is not null && entries.Any(e => lastTrade.TryGetValue(e.Symbol, out var t) && (now - t).TotalSeconds < 300)) return baseSeconds;
        var untilOpen = entries.Select(e => NextOpen(e, now)).OfType<DateTimeOffset>().Select(t => (t - now).TotalSeconds).DefaultIfEmpty(1800).Min();
        return Math.Min(1800, Math.Max(baseSeconds, untilOpen + 5));
    }
}
